edward/inferences/sghmc.py

```diff
@@ -0,0 +1,152 @@
+    friction : float, optional
+      Constant scale on the friction term in the Hamiltonian system.
+    """
+    self.step_size = step_size
+    self.friction = friction
+    self.v = {z: tf.Variable(tf.zeros(qz.params.get_shape()[1:]))
+              for z, qz in six.iteritems(self.latent_vars)}
+    return super(SGHMC, self).initialize(*args, **kwargs)
+
+  def build_update(self):
+    """
+    Simulate Hamiltonian dynamics with friction using a discretized
+    integrator. Its discretization error goes to zero as the learning rate
+    decreases.
+    Implements the update equations from (15) of Chen et al., 2014.
+    """
+    # TODO: Would it be more memory-efficient to use tf.control_dependencies,
+    # rather than repeatedly re-assign attributes stored on self.v?
+    old_sample = {z: tf.gather(qz.params, tf.maximum(self.t - 1, 0))
+                  for z, qz in six.iteritems(self.latent_vars)}
+    old_v_sample = {z: v for z, v in six.iteritems(self.v)}
+
+    # Simulate Hamiltonian dynamics with friction.
+    friction = tf.constant(self.friction, dtype = tf.float32)
+    # TODO: Allow option for exponentially decaying learning rate, or similar.
+    learning_rate = tf.constant(self.step_size * 0.01, dtype = tf.float32)
+    grad_log_joint = tf.gradients(self._log_joint(old_sample),
+                                  list(six.itervalues(old_sample)))
+
+    sample = {}                 # v_sample is so named to indicate that it represents a velocity,
+    v_sample = {}               # rather than a momentum.
+    for z, qz, grad_log_p in \
+        zip(six.iterkeys(self.latent_vars),
+            six.itervalues(self.latent_vars),
+            grad_log_joint):
+      event_shape = qz.get_event_shape()
+      normal = Normal(mu=tf.zeros(event_shape),
+                      sigma=tf.sqrt(learning_rate * friction) * tf.ones(event_shape))
+      sample[z] = old_sample[z] + old_v_sample[z]
+      v_sample[z] = (1. - 0.5 * friction)*old_v_sample[z] \
+                    + learning_rate * grad_log_p + normal.sample()
+
+    # Update Empirical random variables.
+    assign_ops = []
+    variables = {x.name: x for x in
+                 tf.get_default_graph().get_collection(tf.GraphKeys.VARIABLES)}
+    for z, qz in six.iteritems(self.latent_vars):
+      variable = variables[qz.params.op.inputs[0].op.inputs[0].name]
+      assign_ops.append(tf.scatter_update(variable, self.t, sample[z]))
+      assign_ops.append(tf.assign(self.v[z], v_sample[z]).op)
+
+    # Increment n_accept.
+    assign_ops.append(self.n_accept.assign_add(1))
+    return tf.group(*assign_ops)
+
+  def _log_joint(self, z_sample):
+    """
+    Utility function to calculate model's log joint density,
+    log p(x, z), for inputs z (and fixed data x).
+
+    Parameters
+    ----------
+    z_sample : dict
+      Latent variable keys to samples.
+    """
+    # TODO: This appears to be identical across HMC, SGHMC, SGLD. Should it be
```

Comment:
yeah, that sounds useful to refactor (although i'm not sure where the best location is to put it)